harness/determined/layers/_workload_sequencer.py

```diff
@@ -387,6 +387,13 @@ def __iter__(self) -> workload.Stream:
                 yield from self.validate(None)
 
             for op in self.core_context.searcher.operations(core.SearcherMode.ChiefOnly):
+                # This can occur when we have already trained more than our trial
+                # will train for. For example, if we are provided a steps_completed
+                # larger than our searcher max length.
+                if self.batches_until_op_complete(op) <= 0:
+                    logging.info("trial has already completed training")
+                    raise ShouldExit(skip_exit_checkpoint=True)
+
```

Comment:
did you observe a change in behavior when introducing this snippet?

My reading of the code is that without this change:
- `while self.batches_until_op_complete(op) > 0` is skipped entirely
- `yield from self.checkpoint()` will be skipped since checkpoint is already current
- `yield from self.validate(op)` will run if it needs to (which is the correct behavior if you wake up from a checkpoint and find out that you trained but didn't validate)

Reply:
Yeah the difference was is assertion goes off without that line

```
[2023-09-08T18:10:59.136203Z] 3d00dd13 || Traceback (most recent call last):
[2023-09-08T18:10:59.136213Z] 3d00dd13 ||   File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
[2023-09-08T18:10:59.136283Z] 3d00dd13 ||     return _run_code(code, main_globals, None,
[2023-09-08T18:10:59.136286Z] 3d00dd13 ||   File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
[2023-09-08T18:10:59.136343Z] 3d00dd13 ||     exec(code, run_globals)
[2023-09-08T18:10:59.136344Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/harness.py", line 208, in <module>
[2023-09-08T18:10:59.136422Z] 3d00dd13 ||     sys.exit(main(args.train_entrypoint))
[2023-09-08T18:10:59.136423Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/harness.py", line 129, in main
[2023-09-08T18:10:59.136482Z] 3d00dd13 ||     controller.run()
[2023-09-08T18:10:59.136482Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/keras/_tf_keras_trial.py", line 633, in run
[2023-09-08T18:10:59.136608Z] 3d00dd13 ||     self._launch_fit()
[2023-09-08T18:10:59.136608Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/keras/_tf_keras_trial.py", line 668, in _launch_fit
[2023-09-08T18:10:59.136733Z] 3d00dd13 ||     self.model.fit(
[2023-09-08T18:10:59.136734Z] 3d00dd13 ||   File "/opt/conda/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
[2023-09-08T18:10:59.136817Z] 3d00dd13 ||     raise e.with_traceback(filtered_tb) from None
[2023-09-08T18:10:59.136817Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/keras/_tf_keras_trial.py", line 126, in on_train_begin
[2023-09-08T18:10:59.136872Z] 3d00dd13 ||     self.trial_controller._control_loop()
[2023-09-08T18:10:59.136873Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/keras/_tf_keras_trial.py", line 761, in _control_loop
[2023-09-08T18:10:59.137004Z] 3d00dd13 ||     for wkld, response_func in self.workloads:
[2023-09-08T18:10:59.137004Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/layers/_workload_sequencer.py", line 459, in workloads
[2023-09-08T18:10:59.137105Z] 3d00dd13 ||     for wkld, response_fn in wlsq:
[2023-09-08T18:10:59.137106Z] 3d00dd13 ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/layers/_workload_sequencer.py", line 430, in __iter__
[2023-09-08T18:10:59.137199Z] 3d00dd13 ||     assert op._completed, "logic error; op was never completed"
```

would setting ```op._completed = True``` be equivalent?

Reply:
Sounds good to me,

I completely agree on SearcherContext makes this feel pretty complex should

Reply:
This is a great example of why my current project is all about removing SearherContext from the Core API altogether.  It's so confusing that the searcher living in the master requires this tight coupling with the training loop.

Ok, so to summarize, I designed the SearcherContext with two opposing facts in mind:
- Training shall read an op, train some amount, and report it as completed.
- Searcher operations shall be idempotent and based on absolute lengths.

I never realized those were in conflict if you woke up with all your operations done, so this is a pretty severe bug.  I'm not terribly worried about it since it never came up with end users, and we're removing the SearcherContext in the first place.

For now, I'd say replace the `assert op._completed` in the workload sequencer with:

```python
if not op._completed:
    # The only case where op isn't reported as completed is if we restarted but
    # op.length was already trained for and validated on; in that case just break
    # out of the operations loop; we have nothing to do.
    break
```

cc @azhou-determined and @garrett361 who may have an academic interest in this scenario.